app/models/source_resolver.rb

```diff
@@ -0,0 +1,34 @@
+class SourceResolver
+  def self.resolve(new_source)
+    new(new_source).resolve
+  end
+
+  def initialize(new_source)
+    @new_source = new_source
+  end
+
+  def resolve
+    return if @new_source.save
+    adjust_ranks
```

Comment:
Should lines 12 and 13 be wrapped in a transaction so that if something fails everything rolls back?